client/optimizing.go

```diff
@@ -0,0 +1,309 @@
+	// startTime is the time at which the request was started.
+	startTime time.Time
+}
+
+type requestResult struct {
+	// result is the return value from the call to Get.
+	result Result
+	// err is the error that occurred from a call to Get (not including context error).
+	err error
+	// stat is stats from the call to Get.
+	stat *requestStat
+}
+
+func (oc *optimizingClient) testSpeed() {
+	for {
+		stats := []*requestStat{}
+		ctx, cancel := context.WithCancel(context.Background())
+		ch := parallelGet(ctx, oc.clients, 1, oc.requestTimeout, oc.requestConcurrency)
+
+	LOOP:
+		for {
+			select {
+			case rr, ok := <-ch:
+				if !ok {
+					cancel()
+					break LOOP
+				}
+				stats = append(stats, rr.stat)
+			case <-oc.done:
+				cancel()
+				return
+			}
+		}
+
+		oc.updateStats(stats)
+
+		t := time.NewTimer(oc.speedTestInterval)
+		select {
+		case <-t.C:
+		case <-oc.done:
+			t.Stop()
+			return
+		}
+	}
+}
+
+// SetLog configures the client log output.
+func (oc *optimizingClient) SetLog(l log.Logger) {
+	oc.log = l
+}
+
+// Get returns a the randomness at `round` or an error.
+func (oc *optimizingClient) Get(ctx context.Context, round uint64) (res Result, err error) {
+	oc.RLock()
+	// copy the current ordered client list so we iterate over a stable slice
+	var clients []Client
+	for _, s := range oc.stats {
+		clients = append(clients, s.client)
+	}
+	oc.RUnlock()
+
+	stats := []*requestStat{}
+	defer oc.updateStats(stats)
+	ch := raceGet(ctx, clients, round, oc.requestTimeout, oc.requestConcurrency)
+
+LOOP:
+	for {
+		select {
+		case rr, ok := <-ch:
+			if !ok {
+				break LOOP
+			}
+			stats = append(stats, rr.stat)
+			res, err = rr.result, rr.err
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-oc.done:
+			return nil, errors.New("client closed")
+		}
+	}
+
+	return
+}
+
+// get calls Get on the passed client and returns a channel that yields a single
+// raceResult when the call completes or closes the channel with no result if the
+// context is canceled.
+func get(ctx context.Context, client Client, round uint64) <-chan *requestResult {
+	ch := make(chan *requestResult, 1)
+	go func() {
+		start := time.Now()
+		res, err := client.Get(ctx, round)
+		rtt := time.Now().Sub(start)
+
+		if ctx.Err() != nil {
+			close(ch)
+			return
+		}
+
+		// client failure, set a large RTT so it is sent to the back of the list
+		if err != nil {
+			rtt = math.MaxInt64
+		}
+
+		stat := requestStat{client, rtt, start}
+
+		if err != nil {
+			ch <- &requestResult{nil, err, &stat}
+		} else {
+			ch <- &requestResult{res, nil, &stat}
+		}
+		close(ch)
+	}()
+	return ch
+}
+
+func raceGet(ctx context.Context, clients []Client, round uint64, timeout time.Duration, concurrency int) <-chan *requestResult {
+	results := make(chan *requestResult, len(clients))
+
+	go func() {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		defer close(results)
+		ch := parallelGet(ctx, clients, round, timeout, concurrency)
```

Comment:
If I understand correctly, parallelGet queries *all* clients (using fixed concurrency), and it does not stop until it has queried all clients. This is appropriate for testSpeed, but in the codepath Get->raceGet->parallelGet, I think you would want parallel get to terminate as soon as it hears the first answer.

Reply:
This is what is happening below - after a result has been written we check to see if there was an error and if not then we return (which cancels the context passed to parallelGet and closes the channel)